src/java/boa/datagen/slurm/GitHubRepoBareDownloader.java

```diff
@@ -19,62 +20,63 @@ public static void main(String[] args) {
 
 		if (args.length < 3) {
 			System.out.println("args: INPUT_NAMES_PATH, OUTPUT_REPOS_PATH, THREAD_NUM");
-		} else {
-			INPUT_PATH = args[0];
-			OUTPUT_REPOS_PATH = args[1];
-			THREAD_NUM = Integer.parseInt(args[2]);
-
-			File input = new File(INPUT_PATH);
-
-			DownloadWorker[] workers = new DownloadWorker[THREAD_NUM];
-			Thread[] threads = new Thread[THREAD_NUM];
-			for (int i = 0; i < THREAD_NUM; i++) {
-				workers[i] = new DownloadWorker(i);
-				threads[i] = new Thread(workers[i]);
-				threads[i].start();
-			}
+			return;
+		}
 
-			// assign tasks to workers
-			for (File file : input.listFiles()) {
-				if (!file.getName().endsWith(".json"))
-					continue;
-				JsonElement jsonTree = new JsonParser().parse(FileIO.readFileContents(file));
-				for (JsonElement je : jsonTree.getAsJsonArray()) {
-					String projectName = je.getAsJsonObject().get("html_url").getAsString()
-							.replace("https://github.com/", "");
-
-					boolean assigned = false;
-					while (!assigned) {
-						for (int j = 0; j < THREAD_NUM; j++) {
-							if (workers[j].isReady()) {
-								workers[j].setName(projectName);
-								workers[j].setReady(false);
-								assigned = true;
-								break;
-							}
-						}
-						try {
-							Thread.sleep(100);
-						} catch (InterruptedException e) {
-							e.printStackTrace();
+		INPUT_PATH = args[0];
+		OUTPUT_REPOS_PATH = args[1];
+		THREAD_NUM = Integer.parseInt(args[2]);
+
+		File input = new File(INPUT_PATH);
+
+		DownloadWorker[] workers = new DownloadWorker[THREAD_NUM];
+		Thread[] threads = new Thread[THREAD_NUM];
+		for (int i = 0; i < THREAD_NUM; i++) {
+			workers[i] = new DownloadWorker(i);
+			threads[i] = new Thread(workers[i]);
+			threads[i].start();
+		}
+
+		// assign tasks to workers
+		for (File file : input.listFiles()) {
+			if (!file.getName().endsWith(".json"))
+				continue;
+			JsonElement jsonTree = new JsonParser().parse(FileIO.readFileContents(file));
+			for (JsonElement je : jsonTree.getAsJsonArray()) {
+				String projectName = je.getAsJsonObject().get("html_url").getAsString().replace("https://github.com/",
+						"");
+
+				boolean assigned = false;
+				while (!assigned) {
+					for (int j = 0; j < THREAD_NUM; j++) {
```

Comment:
If Boa code is intended to compile using Java 7 or higher, one could use concurrency features such as ExecutorService that is meant to provide all of these features. Just a thought. See https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ThreadPoolExecutor.html